File: haskell/p05.hs
-- P05 Reverse a list.

module P05 where

-- Predefined function
f0 :: [a] -> [a]
f0 = reverse

-- Recursion (inefficient)
f1 :: [a] -> [a]
f1 [] = []
f1 (x:xs) = f1 xs ++ [x]

-- Tail recursion
f2 :: [a] -> [a]
f2 = f2' []
 where f2' :: [a] -> [a] -> [a]
       f2' acc [] = acc
       f2' acc (x:xs) = f2' (x:acc) xs

-- Folding
f3 :: [a] -> [a]
f3 = foldl (flip (:)) []

File: scala/P05.scala
// P05 Reverse a list.

// Predefined function
def f0[T](list: List[T]): List[T] = list.reverse

// Recursion (inefficient)
def f1[T]: List[T] => List[T] = {
  case Nil => Nil
  case h :: t => f1(t) ::: h :: Nil
}

// Tail recursion
@annotation.tailrec
def f2[T](list: List[T], acc: List[T] = Nil): List[T] = list match {
  case Nil => acc
  case h :: t => f2(t, h :: acc)
}

// Folding
def f3[T](list: List[T]): List[T] =
  list.foldLeft(Nil: List[T])((a, b) => b :: a)

// Imperative
def f4[T](list: List[T]): List[T] = {
  var acc: List[T] = Nil
  for(each <- list) acc ::= each
  acc
}